### ISSUE TITLE
HTML5 Export >40MB

### ISSUE THREAD

**Issue by netAction** [godotengine/godot#68647]
### Godot version

4.0 Beta 4

### System information

Windows 64

### Issue description

The HTML5 export of a single empty 2D node has a size of >40MB. Godot 3.5 HTML5 export is already >20MB.

Users on low bandwidth are kept away from the game. The most important advantage of HTML5 over native apps is the easy access. You can start a game via scanning a QR code. But exactly this benefit of HTML5 is weakened in regions like Germany where cellular coverage and wifi quality are low.

The other issue is the CPU load caused by the huge amount of executable code. 40MB JS and WASM code affects the device speed more than a PNG of 40MB. I didn't benchmark this though.

I am aware Godot does come with some basic functionality and WASM wrapper that already needs about 1MB. But the other engine size could be optimized. Maybe a selection of different engines could make sense that are chosen depending on the featured needed for the game.

There are methods for compression of WASM. It can be done by Apache mod_deflate or even the Javascript zlib library pako. This helps a lot but does not shrink the game by a magnitude.

### Steps to reproduce

Create a new project. Add 2D node. Export HTML5.

**Comment by Zireael07** [comment]
Related: https://github.com/godotengine/godot/issues/41118 (the first comment mentions the biggest factor is likely the sheer size of the WASM payload)

Related: https://github.com/godotengine/godot-proposals/issues/190 which would be likely closed when https://github.com/godotengine/godot/pull/47054 is merged

**Comment by Calinou** [comment]
> Maybe a selection of different engines could make sense that are chosen depending on the featured needed for the game.

This is now implemented with [feature build profiles](https://github.com/godotengine/godot/pull/62996), providing you build a custom HTML5 export template.

**Comment by clayjohn** [comment]
It would be helpful for us to investigate where exactly the extra 20 mbs are coming from. While the size of the engine has certainly increased, it hasn't doubled in size. 

Perhaps we are mistakenly including debug symbols or something, or a dependency that wasn't previously exposed to HTML5 builds now is. 

On that note, @netAction Could you confirm whether you are using debug or release builds for the HTML5 export?

**Comment by netAction** [comment]
> Could you confirm whether you are using debug or release builds for the HTML5 export?

After switching off Debug in Godot 4.0 beta4, the export has 24MB rather than 41MB.

After switching off Debug in Godot 3.5, the export has 13MB rather than 17MB.

Sorry I missed this option.

**Comment by rthery** [comment]
I'm investigating if Godot could replace my engine of choice for web development (I'd prefer to code in C# than TS) and thought I'd share some insights.
The current size of the engine to download is indeed a show stopper for now. The initial download time is crucial in terms of acquisition, so at the moment Godot is not viable professionally. (Which is the same problem with Unity, they tried to solve via their cancelled Tiny initiative, it's a very hard problem to solve for C++ engines)

I just tested Godot 4.0 beta8, I got these results with a release web build with just a cube, a light and a camera
26M     Godot.wasm
6.6M    Godot.wasm.gz (compressed with gzip)
4.4M    Godot.wasm.br (compressed with brotli)

Aiming for something around 1-2M after compression would still be worst than JS based engines (I don't think Godot will ever be able to beat those), but competitive enough to be used professionally for web games :)

Are the feature build profiles already available somewhere in the beta ? I don't understand how to get that window displayed in the PR.

**Comment by Calinou** [comment]
> Are the feature build profiles already available somewhere in the beta ? I don't understand how to get that window displayed in the PR.

Yes, but you need to compile the export template from source to benefit from the generated build profile.

**Comment by miguelportilla** [comment]
Please excuse my ignorance, I am new to Godot but how do you export to HTML5 with Godot 4? I am using the beta 8 version. 

I've downloaded and inspected the export templates zip file and I only see Mac, Linux, and Windows templates.

Thanks

**Comment by akien-mga** [comment]
You must be using the .NET build, which currently only supports desktop platforms (like .NET Core itself).

**Comment by indubitablement2** [comment]
Could something like Defold's build server help by allowing less experienced users remove parts of the engine they don't use?

https://defold.com/manuals/extensions/#the-build-server

They are emscripten based as well and are hable to get down to <1MB for simple game like pong.

**Comment by slabgames** [comment]
is there an improvement in this area?

**Comment by Calinou** [comment]
> is there an improvement in this area?

There are a few PRs that aim to address this like https://github.com/godotengine/godot/pull/101964, but they will require custom builds as they come with compromises on performance.
